src/wp-includes/general-template.php

```diff
@@ -3413,6 +3413,107 @@ function wp_resource_hints() {
 	}
 }
 
+/**
+ * Prints resource preloads directives to browsers.
+ * https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload
+ *
+ * Gives directive to browsers to preload specific resources that website will
+ * need very soon, this ensures that they are available earlier and are less
+ * likely to block the page's render.
+ *
+ * These performance improving indicators work by using `<link rel"preload">`.
+ *
+ * @since 6.x.x
+ */
+function wp_preload_links() {
+	/**
+	 * Filters domains and URLs for resource preloads.
+	 *
+	 * @since 6.x.x
+	 *
+	 * @param array  $urls {
+	 *     Array of resources and their attributes, or URLs to print for resource preloads.
+	 *
+	 *     @type array ...$0 {
+	 *         Array of resource attributes.
+	 *
+	 *         @type string $href        URL to include in resource preloads. Required.
+	 *         @type string $as          How the browser should treat the resource
+	 *                                   (`script`, `style`, `image`, `document`, etc).
+	 *         @type string $crossorigin Indicates the CORS policy of the specified resource.
+	 *         @type string $type        Type of the resource (`text/html`, `text/css`, etc).
+	 *         @type string $media       Accepts media types or media queries. Allows responsive preloading.
+	 *     }
+	 * }
+	 */
+	$urls = apply_filters( 'wp_preload_links', array() );
+
+	if ( ! is_array( $urls ) ) {
+		return;
+	}
+
+	$unique_urls = array();
+	foreach ( $urls as $url ) {
+		if ( ! is_array( $url ) ) {
+			continue;
+		}
+
+		$atts = $url;
+		if ( isset( $url['href'] ) ) {
+			$url          = esc_url( $url['href'], array( 'http', 'https' ) );
+			$atts['href'] = $url;
+			if ( isset( $unique_urls[ $url ] ) ) {
+				continue;
+			}
+			$unique_urls[ $url ] = $atts;
+			// Media can use imagesrcset and not href.
+		} elseif ( ! isset( $url['href'] ) && ( 'image' === $url['as'] ) && ( isset( $url['imagesrcset'] ) || isset( $url['imagesizes'] ) ) ) {
+			if ( isset( $unique_urls[ $url['imagesrcset'] ] ) ) {
+				continue;
+			}
+			$unique_urls[ $url['imagesrcset'] ] = $atts;
+		} else {
+			continue;
+		}
+
+		foreach ( $unique_urls as $atts ) {
+			$html = '';
+
+			foreach ( $atts as $attr => $value ) {
+				if ( ! is_scalar( $value )
+					 || ( ! in_array( $attr, array( 'as', 'crossorigin', 'href', 'imagesrcset', 'imagesizes', 'type', 'media' ), true ) && ! is_numeric( $attr ) )
+				) {
```

Comment:
Nit: splitting this across multiple lines makes it a bit more readable. Right now the line is a bit long.